### ISSUE TITLE
[docs] [feature request] [sdpa] Minor docs suggestions on F.scaled_dot_product_attention + proposal to add explicit argument for backend selection

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#96968]
### 📚 The doc issue

https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html?highlight=scaled_dot_product_attention#torch.nn.functional.scaled_dot_product_attention:

The code example is rather not ordinary and contains a `return` statement outside of any function definition:
<img width="455" alt="image" src="https://user-images.githubusercontent.com/1041752/225709394-f79f1ec5-4b75-4e50-b129-8db3cd71f39a.png">

is `not attn_mask` correct? should it not be `~attn_mask` instead or at least `attn_mask.logical_not()`? Also, https://pytorch.org/docs/stable/generated/torch.logical_not.html does not say if it correspons to `~` or to `not`

It would be better to add some `def reference_scaled_dot_product_attention(...):` to this code example, then it would be copy-pasteable for education purpose

<img width="282" alt="image" src="https://user-images.githubusercontent.com/1041752/225709769-a99f30d9-0021-44a5-9f6c-1eadbb6d9e3f.png">

`enable_math` argname is also quite not self-explanatory :) and the comment `Enables or Disables the PyTorch C++ implementation.` doesn't match/explain much.

<img width="504" alt="image" src="https://user-images.githubusercontent.com/1041752/225712873-b67f0699-695e-4c2b-90c0-e3b28c7a7053.png">

from the blog post https://pytorch.org/blog/pytorch-2.0-release/ also does not explain what `math` mean in this context (actually, same question about mem-efficient attention. it would be great if this doc page included an overview of properties of different attention backends)

`Memory-efficient attention` link only points to https://github.com/facebookresearch/xformers without any specific README section or code file. Also, if xformers is mentioned wrt mem_efficient_attention, it should be explained how xformers repo is currently related to core pytorch implementations

An unrelated feature request: support directly `backend`/similar argument to force the SDPA implementation explicitly without context managers (more general discussion here: https://github.com/pytorch/pytorch/issues/52439)

**Comment by mikaylagawarecki** [comment]
@drisspg

**Comment by drisspg** [comment]
Thanks for pointing these out, I will incorporate some of them!

**Comment by mikekgfb** [comment]
There is only a single interface to scaled dot product attention in PyTorch, scaled_dot_product_attention. The other implementations are kernels that reflect optimizations for specific hardware SKUs, and may reflect a variable set of constraints on the inputs.  Because they are not part of the committed interface of Pytorch, the constraints may change, or the function may be removed at any time.

**Comment by vadimkantorov** [comment]
I see what you mean :) so maybe it's best to move these implementation details to some special dev doc page, my point is that they should be there somewhere for power users eager to learn and find what works best on their compute (this will happen especially at the early stage, advanced users are going to compare it with evolving xformers and flash attention codebase a, so it's important to know what is actually implemented in core and whether spending time on experimentation is worth or not)

**Comment by mikekgfb** [comment]
> I see what you mean :) so maybe it's best to move these implementation details to some special dev doc page, my point is that they should be there somewhere for power users eager to learn and find what works best on their compute (this will happen especially at the early stage, advanced users are going to compare it with evolving xformers and flash attention codebase a, so it's important to know what is actually implemented in core and whether spending time on experimentation is worth or not)

We understand the desire to know more about the implementation which is why we published this blog to add more context and information => https://pytorch.org/blog/accelerated-pytorch-2/

For the most advanced users, the details may be derived from the source code which is available, as well as from the blog we have published. Documenting the limitations as part of the interface might be misleading, because it might be construed as representing an interface commitment.

We can't even guarantee the bounds in which these kernels work, because they are SKU dependent.  We resisted SKU-level optimization in Pytorch for the longest time, because it introduces several significant issuesabout compatibility between SKUs, support, testing etc - see also the discussion surrounding support for these functions on the 3090 and 4090 SKUs.

The performance opportunity was just too big and so we wanted to make this capability available to our users.  And -- we have done exactly what you suggested, gathered a description of the boundaries for different kernels separatelky, as a blog here => https://pytorch.org/blog/accelerated-pytorch-2/
